src/FffGcodeWriter.cpp

```diff
@@ -1487,38 +1492,107 @@ bool FffGcodeWriter::addSupportRoofsToGCode(const SliceDataStorage& storage, Lay
+        fill_angle = 90; // perpendicular to support lines
     }
     else 
     {
-        fillAngle = 45 + (((layer_nr % 2) + 2) % 2) * 90; // alternate between the two kinds of diagonal:  / and \ .
+        fill_angle = 45 + (((layer_nr % 2) + 2) % 2) * 90; // alternate between the two kinds of diagonal:  / and \ .
         // +2) %2 to handle negative layer numbers
     }
-    int support_skin_overlap = 0; // the skin (roofs/bottoms) should never be expanded outwards
-    int outline_offset =  0;
-    int extra_infill_shift = 0;
-    Polygons* perimeter_gaps = nullptr;
-    bool use_endpieces = true;
-    bool connected_zigzags = false;
-    Infill infill_comp(pattern, support_layer.skin, outline_offset, gcode_layer.configs_storage.support_interface_config.getLineWidth(), support_line_distance, support_skin_overlap, fillAngle, z, extra_infill_shift, perimeter_gaps, connected_zigzags, use_endpieces);
-    Polygons support_polygons;
-    Polygons support_lines;
-    infill_comp.generate(support_polygons, support_lines);
-
-    if (support_lines.size() > 0 || support_polygons.size() > 0)
-    {
-        setExtruder_addPrime(storage, gcode_layer, layer_nr, skin_extruder_nr);
-        gcode_layer.addPolygonsByOptimizer(support_polygons, &gcode_layer.configs_storage.support_interface_config);
-        gcode_layer.addLinesByOptimizer(support_lines, &gcode_layer.configs_storage.support_interface_config, (pattern == EFillMethod::ZIG_ZAG)? SpaceFillType::PolyLines : SpaceFillType::Lines);
+    constexpr int support_roof_overlap = 0; // the roofs should never be expanded outwards
+    constexpr int outline_offset =  0;
+    constexpr int extra_infill_shift = 0;
+    constexpr Polygons* perimeter_gaps = nullptr;
+    constexpr bool use_endpieces = true;
+    constexpr bool connected_zigzags = false;
+
+    const coord_t support_roof_line_distance = roof_extr.getSettingInMicrons("support_roof_line_distance");
+    Infill roof_computation(pattern, support_layer.support_roof, outline_offset, gcode_layer.configs_storage.support_roof_config.getLineWidth(), support_roof_line_distance, support_roof_overlap, fill_angle, z, extra_infill_shift, perimeter_gaps, connected_zigzags, use_endpieces);
+    Polygons roof_polygons;
+    Polygons roof_lines;
+    roof_computation.generate(roof_polygons, roof_lines);
+    if (!roof_polygons.empty() || !roof_lines.empty())
+    {
+        setExtruder_addPrime(storage, gcode_layer, layer_nr, roof_extruder_nr);
+        gcode_layer.addPolygonsByOptimizer(roof_polygons, &gcode_layer.configs_storage.support_roof_config);
+        gcode_layer.addLinesByOptimizer(roof_lines, &gcode_layer.configs_storage.support_roof_config, (pattern == EFillMethod::ZIG_ZAG) ? SpaceFillType::PolyLines : SpaceFillType::Lines);
+        added = true;
+    }
+
+    return added;
+}
+
+bool FffGcodeWriter::addSupportBottomsToGCode(const SliceDataStorage& storage, LayerPlan& gcode_layer, int layer_nr) const
```

Comment:
I removed some duplication and made the code more compact so that there are now no two consecutive statements that aren't specific to roof/bottom.